### PR TITLE
adding debugging guide

### PR DESCRIPTION
**Pull request by heckj**
- added in sequence diagram to explain default (discovery) workflow
- added in logs and details of current code to verify early states in the
  sequence

### CODE REVIEW

File: docs/devguide/debugging_guide.rst [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
...cmVhZHkgImtub3duIiwgaXQgd2lsbCBvbmwAg0oIZCBpADAFcmUncyBhbiBhY3RpdmUgdwCEYgcgdGhhdCdzIGJlZW4gaW52b2tlZCByZWxhdGVkIHRvAIEeCQBsJW4ndCAAgQYFAIEACmNyZWF0ZSBhAGcKKGQAhW8HAHsIKQCDQAUAgSYIAIUmBWFuIGkAgiYOAINcCACBKwUAhWERAIQGBQCDUwcoAINNBQCDQAdjYWxscyBhIFByb2ZpbGUpICh2aWEgAINRCwCEWxAAPgwAhjEFZACHEwltaWNyb2tlcm5lbACEbglyZACEIggAhyAQAIUZBQCEXQlzdGF0aWMAhW0GLQCDdwV2bWxpbnV6IABLBgCHMxEAFwgAKzMAgRIGAIISEgCBIg0AhkYidGhlAIEWBwCBYgxhbmQgdHJhbnNmZXJzIGNvbnRyb2wgKGJvb3RzAIRHBQCCGwwAhzgXdGhlAIJAF2lzIHNlAIhYBQCGOAhhbmQgbGF1bmNoIGEgTm9kZUpTIHRhc2sgcnVubm4Ah1ITAIdQCXRoZQCJQgVzdHJhcC5qcyB0ZW1wbGF0ZQCKHxEAGQ1maWxsZWQgb3UAg3kHdmFsdWVzIHNwZWNpZmljAIYIDCBiYXNlZCBvbiBhAIgHBXVwAIkpFnJ1bnMAKgcAgQULAItREQCBJAkgYXNrcyBmb3IAgWUFcwCFSwdzaG91bGQgSSBkbz8Ai2ISZGF0YSBwYWNrZQCKNQUAMgcAhU4fAIxmEiBwYXNzZXMAi1cFAHcHAIZ4BXRlcnJvZ2F0ZSBoYXJkd2FyZQpsb29wAIEgBWVhY2ggVGFzawCKNQwKICAgAIs5BwCNOgpvdXRwdQCBFwkKZW5kAIpKFwCDXQUAJwdzdG9yZWQgYXMgY2F0YWxvZ3MgaW4Aix8IAIkAKmkAizcIIGkAhTMFZmlndXJlAIhpB1NLVSBkZWYAjEQFb25zAIpEBXByb2NlAIIIBXRoZXNlAHQKdG8gZGV0ZXJtaW5lAItCBVNLVQBbGgCKcgkAYAUAhD4JAIp3CQByBWVkLACGQwkAgRkGdGludQCKeghhdCwgb3RoZXJ3aXMAbAYAjDQFAItlBWJlIHRvbACLJgVyZWJvb3QAhAlOTm90aGluZyBtb3JlLCB0aGFua3MgLSBwbGVhc2UAawcAikgMCgo&s=napkin
```
Comment (benbp) [inline docs/devguide/debugging_guide.rst]: Is there any way you can output a higher resolution source image? The original is kind of hard to read as it is.

File: docs/devguide/debugging_guide.rst [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    title Default Discovery Workflow
+    Server->RackHD: DHCP from PXE(nic or BIOS)
+    RackHD->Server: ISC DHCP response with IP
+    RackHD->Server: DHCP-proxy response with bootfile
+    Server->RackHD: Request to download bootfile via TFTP
+    RackHD->Server: TFTP sends requested file (monorail.ipxe)
+    note left of Server: Server loads monorail.ipxe and initiates on bootloader
+    Server->RackHD: IPXE script requests what to do from RackHD (http)
+    note right of RackHD: RackHD looks up IP address of HTTP request from IPXE script to find the node via it's mac-address.
+    note right of RackHD: If the node is already "known", it will only respond if there's an active workflow that's been invoked related to the node
```
Comment (benbp) [inline docs/devguide/debugging_guide.rst]: This step actually happens before we even respond with the first DHCP-proxy packet containing the bootfile. This is why we see the "Did not receive any proxyDHCP offers" message.

File: docs/devguide/debugging_guide.rst [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    note right of RackHD: If the node isn't known, it will create a workflow (default workflow) and respond with an iPXE script to initiate that
```
Comment (benbp) [inline docs/devguide/debugging_guide.rst]: Perhaps "default discovery workflow".

File: docs/devguide/debugging_guide.rst [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    RackHD->Server: ipxe script (what RackHD calls a Profile) (via http)
+    note left of Server: ipxe script with discovery microkernel and initrd (http)
+    Server->RackHD: ipxe requests static file - the vmlinuz kernel
+    RackHD->Server: vmlinuz (http)
+    Server->RackHD: ipxe requests static file - initrd
+    RackHD->Server: initrd (http)
+    note left of Server: Server loads the kernel and initrd and transfers control (boots that microkernel)
```
Comment (benbp) [inline docs/devguide/debugging_guide.rst]: Might also be helpful to include another step here: The initrd runs custom scripts that download the microkernel image and boot it from RAM.

File: docs/devguide/debugging_guide.rst (hunk 3 above)
Comment (heckj) [inline docs/devguide/debugging_guide.rst]: updated

File: docs/devguide/debugging_guide.rst (hunk 4 above)
Comment (heckj) [inline docs/devguide/debugging_guide.rst]: updated

File: docs/devguide/debugging_guide.rst (hunk 1 above)
Comment (heckj) [inline docs/devguide/debugging_guide.rst]: can't really - not and use what's available, this is a dynamic link to public site (websequencediagrams)

**Comment by heckj** [comment]
I fiddled with some variations on the sequence diagrams - if you constrict what's in the diagram, it gets more readable, but there doesn't appear to be size controls for the resulting image

**Comment by benbp** [comment]
:+1:

File: docs/devguide/debugging_guide.rst [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    title Default Discovery Workflow
+    Server->RackHD: DHCP from PXE(nic or BIOS)
+    RackHD->Server: ISC DHCP response with IP
+    RackHD->Server: DHCP-proxy response with bootfile
+    note right of RackHD: If the node is already "known", it will only respond if there's an active workflow that's been invoked related to the node
+    Server->RackHD: Request to download bootfile via TFTP
+    RackHD->Server: TFTP sends requested file (monorail.ipxe)
+    note left of Server: Server loads monorail.ipxe and initiates on bootloader
+    Server->RackHD: IPXE script requests what to do from RackHD (http)
+    note right of RackHD: RackHD looks up IP address of HTTP request from IPXE script to find the node via it's mac-address.
```
Comment (yyscamper) [inline docs/devguide/debugging_guide.rst]: typo at the end of line, should be **its**, not **it's**

File: docs/devguide/debugging_guide.rst [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    note right of RackHD: If the node is already "known", it will only respond if there's an active workflow that's been invoked related to the node
+    note right of RackHD: If the node isn't known, it will create a workflow (default is a discovery workflow) and respond with an iPXE script to initiate that
```
Comment (yyscamper) [inline docs/devguide/debugging_guide.rst]: Strictly speaking, it should be "SKU discovery workflow" (Graph.Sku.Discovery, https://github.com/RackHD/on-http/blob/master/lib%2Fservices%2Fprofiles-api-service.js#L123), the discovery workflow (Graph.Discovery) is nested in Graph.Sku.Discovery.

File: docs/devguide/debugging_guide.rst [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    RackHD->Server: bootstrap.js filled out with values specific to the node based on a lookup
+    note left of Server: runs node bootstrap.js
+    Server->RackHD: bootstrap asks for tasks (what should I do?)
+    RackHD->Server: data packet of tasks (via http)
+    note left of Server: Discovery Workflow passes down tasks to interrogate hardware
+    loop for each Task from RackHD
+        Server->RackHD: output of task
+    end
+    note right of RackHD: task output stored as catalogs in RackHD related to the node
+    note right of RackHD: if RackHD is configured with SKU definitions, it processes these catalogs to determine the SKU
```
Comment (yyscamper) [inline docs/devguide/debugging_guide.rst]: After SKU is determined, RackHD will continue do following work [discovery-sku-graph.js](https://github.com/RackHD/on-taskgraph/blob/master/lib/graphs/discovery-sku-graph.js):
- Generate enclosure nodes bases on catalog data.
- Create default pollers.

**Comment by yyscamper** [comment]
It will be better to correct the iPXE spelling, sometimes you write as "IPXE", sometimes as "ipxe" or "iPXE". The "iPXE" should be the correct one as official site uses it.

File: docs/devguide/debugging_guide.rst [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    Server->RackHD: requests the bootstrap.js template
+    RackHD->Server: bootstrap.js filled out with values specific to the node based on a lookup
+    note left of Server: runs node bootstrap.js
+    Server->RackHD: bootstrap asks for tasks (what should I do?)
+    RackHD->Server: data packet of tasks (via http)
+    note left of Server: Discovery Workflow passes down tasks to interrogate hardware
+    loop for each Task from RackHD
+        Server->RackHD: output of task
+    end
+    note right of RackHD: task output stored as catalogs in RackHD related to the node
```
Comment (yyscamper) [inline docs/devguide/debugging_guide.rst]: Whether to speak of the "task output parser", RackHD will parse the task output and store the catalog as JSON format.

File: docs/devguide/debugging_guide.rst (hunk 5 above)
Comment (heckj) [inline docs/devguide/debugging_guide.rst]: got it, thanks!

File: docs/devguide/debugging_guide.rst (hunk 8 above)
Comment (heckj) [inline docs/devguide/debugging_guide.rst]: wasn't going to try and go into all the detail here, mostly aiming at a clean sequence diagram of the major interactions between systems., so didn't mention it

**Comment by yyscamper** [comment]
:+1:

**Comment by yyscamper** [comment]
@heckj : Whether can you generate a higher resolution picture? and using the regular printing font, it is a little difficulty for me (and some other Shanghai people) to understand the  handwritten font :(

**Pull request merged**